pegasus/helpers/form_helpers.rb

```diff
@@ -56,25 +56,25 @@ def insert_or_upsert_form(kind, data, options={})
 
   # For HOC signups, a duplicate (matching email and name) entry is assumed to be an update rather
   # than an insert.
-  num_rows_updated = 0
-  if (kind =~ /HocSignup/) == 0
+  if kind.start_with? 'HocSignup'
+    row_for_update = row.dup
+    [:created_at, :created_ip, :secret].each {|key| row_for_update.delete key}
     num_rows_updated = DB[:forms].
-      where(email: row[:email], kind: kind, name: row[:name]).
-      update(row)
+      where(kind: kind, email: row_for_update[:email], name: row_for_update[:name]).
+      update(row_for_update)
+    return row_for_update if num_rows_updated > 0
```

Comment:
Aside, I don't like the code (data manipulation) duplication between `insert_and_upsert_form` and `update_form`, nor the structure of `insert_and_upsert_form`. That methods don't return what you want them to return will make the refactor slightly more complicated.